### ISSUE TITLE
DRAM-sharded matmul hanging when invoked twice

### ISSUE THREAD

**Issue by barci2** [tenstorrent/tt-metal#24681]
### Describe the bug

During the MLP bringup for DeepSeek we noticed that, when running in decode mode, the module would not produce a result and would just hang for over 10 minutes after the execution was killed. We managed to narrow the issue down to the second run of a matmul hanging the execution. If running in slow dispatch mode, this issue appears once we run the op. If ran in the fast dispatch mode (the default), this becomes noticeable once we try to print the contents of the tensor (i.e. when it has to be read back). This also happens on a single-card execution.

### Steps to reproduce the issue

Run the script below in the environment build from main:
```python
import torch
import ttnn
from ttnn.types import TILE_SIZE

COMPUTE_KERNEL_CONFIG_LOFI = ttnn.WormholeComputeKernelConfig(
    math_fidelity=ttnn.MathFidelity.LoFi,
    math_approx_mode=False,
    fp32_dest_acc_en=False,
    packer_l1_acc=True,
)


def dram_sharded_weight_config(k, n, dram_grid_size):
    """Create DRAM-sharded memory config for width-sharded tensors"""
    dram_cores = dram_grid_size.x  # WH has 12 dram cores, P150 has 8, P100 has 7
    assert dram_grid_size.y == 1, "Current dram sharding assumes y dim is 1"
    dram_weight_grid = ttnn.CoreRangeSet(
        {
            ttnn.CoreRange(
                ttnn.CoreCoord(0, 0),
                ttnn.CoreCoord(dram_grid_size.x - 1, dram_grid_size.y - 1),
            )
        }
    )

    shard_spec = ttnn.ShardSpec(
        dram_weight_grid,
        (k, ttnn.core.roundup(ttnn.core.divup(n, dram_cores), TILE_SIZE)),
        ttnn.ShardOrientation.ROW_MAJOR,
    )
    return ttnn.MemoryConfig(ttnn.TensorMemoryLayout.WIDTH_SHARDED, ttnn.BufferType.DRAM, shard_spec)


def report_tensor(tt_tensor, name):
    for e in [
        ".buffer_address()",
        ".device()",
        ".dtype",
        ".is_allocated()",
        ".is_sharded()",
        ".layout",
        ".logical_volume()",
        ".memory_config()",
        ".padded_shape",
        ".storage_type()",
        ".tensor_id",
        ".volume()",
    ]:
        print(f"{name}{e} = {eval('tt_tensor' + e)}")
    print(f"{name} = {tt_tensor}")


def big_print(msg):
    """Print a message with a big border."""
    border = "=" * (len(msg) + 4)
    print(f"\n{border}\n  {msg}\n{border}\n\n")


def test_forward_pass():
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 1))

    activation_sharding_num_cores = 7
    batch_size = 32
    dim = 7168
    hidden_dim = 18432

    input_memory_config = ttnn.MemoryConfig(
        memory_layout=ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        buffer_type=ttnn.BufferType.L1,
        shard_spec=ttnn.ShardSpec(
            grid=ttnn.num_cores_to_corerangeset(
                activation_sharding_num_cores, (mesh_device.core_grid.x, mesh_device.core_grid.y)
            ),
            shard_shape=(
                ttnn.core.roundup(batch_size, TILE_SIZE),
                ttnn.core.roundup(ttnn.core.divup(dim, activation_sharding_num_cores), TILE_SIZE),
            ),
            shard_orientation=ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )
    print(input_memory_config)

    w1 = ttnn.from_torch(
        torch.rand((1, 1, dim, hidden_dim)),
        dtype=ttnn.bfloat4_b,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        memory_config=dram_sharded_weight_config(
            dim,
            hidden_dim,
            mesh_device.dram_grid_size(),
        ),
        mesh_mapper=ttnn.ttnn.shard_tensor_to_mesh_mapper(mesh_device, 1),
    )
    w2 = ttnn.from_torch(
        torch.rand((1, 1, hidden_dim, dim)),
        dtype=ttnn.bfloat4_b,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        memory_config=dram_sharded_weight_config(
            hidden_dim,
            dim,
            mesh_device.dram_grid_size(),
        ),
        mesh_mapper=ttnn.ttnn.shard_tensor_to_mesh_mapper(mesh_device, 0),
    )
    tt_input = ttnn.from_torch(
        torch.randn(1, 1, batch_size, dim),
        device=mesh_device,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
        dtype=ttnn.bfloat16,
        memory_config=input_memory_config,
        layout=ttnn.TILE_LAYOUT,
    )

    # Run the ops
    big_print("w1_out = ttnn.linear(tt_input, **cfg['w1'])")
    w1_out = ttnn.linear(
        tt_input,
        input_tensor_b=w1,
        memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG,
        program_config=ttnn.MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig(
            in0_block_w=1,
            per_core_M=1,
            per_core_N=ttnn.core.divup(hidden_dim, TILE_SIZE * activation_sharding_num_cores),
            fused_activation=None,
        ),
        compute_kernel_config=COMPUTE_KERNEL_CONFIG_LOFI,
    )
    report_tensor(w1_out, "w1_out")

    big_print("tt_output = ttnn.linear(w1_out, **cfg['w2'])")
    tt_output = ttnn.linear(
        w1_out,
        input_tensor_b=w2,
        memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG,
        program_config=ttnn.MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig(
            in0_block_w=1,
            per_core_M=1,
            per_core_N=ttnn.core.divup(dim, TILE_SIZE * activation_sharding_num_cores),
            fused_activation=None,
        ),
        compute_kernel_config=COMPUTE_KERNEL_CONFIG_LOFI,
    )
    report_tensor(tt_output, "tt_output")
    big_print("FINISHED EXECUTION")


test_forward_pass()
```

### Expected behavior

The script finishes execution within at most 15 minutes.

### Please complete the following environment information

system: T3K (IRD)
docker: latest build

**Comment by yieldthought** [comment]
This is affecting DeepSeek 671B bringup, we'd very much like to know what's wrong here.

**Comment by barci2** [comment]
It's worth noting here that a core count 32 doesn't hang

**Comment by barci2** [comment]
Smaller example here (only one OP)
```python
import torch
import ttnn
from ttnn.types import TILE_SIZE

COMPUTE_KERNEL_CONFIG_LOFI = ttnn.WormholeComputeKernelConfig(
    math_fidelity=ttnn.MathFidelity.LoFi,
    math_approx_mode=False,
    fp32_dest_acc_en=False,
    packer_l1_acc=True,
)


def dram_sharded_weight_config(k, n, dram_grid_size):
    """Create DRAM-sharded memory config for width-sharded tensors"""
    dram_cores = dram_grid_size.x  # WH has 12 dram cores, P150 has 8, P100 has 7
    assert dram_grid_size.y == 1, "Current dram sharding assumes y dim is 1"
    dram_weight_grid = ttnn.CoreRangeSet(
        {
            ttnn.CoreRange(
                ttnn.CoreCoord(0, 0),
                ttnn.CoreCoord(dram_grid_size.x - 1, dram_grid_size.y - 1),
            )
        }
    )

    shard_spec = ttnn.ShardSpec(
        dram_weight_grid,
        (k, ttnn.core.roundup(ttnn.core.divup(n, dram_cores), TILE_SIZE)),
        ttnn.ShardOrientation.ROW_MAJOR,
    )
    return ttnn.MemoryConfig(ttnn.TensorMemoryLayout.WIDTH_SHARDED, ttnn.BufferType.DRAM, shard_spec)


def report_tensor(tt_tensor, name):
    for e in [
        ".buffer_address()",
        ".device()",
        ".dtype",
        ".is_allocated()",
        ".is_sharded()",
        ".layout",
        ".logical_volume()",
        ".memory_config()",
        ".padded_shape",
        ".storage_type()",
        ".tensor_id",
        ".volume()",
    ]:
        print(f"{name}{e} = {eval('tt_tensor' + e)}")
    print(f"{name} = {tt_tensor}")


def big_print(msg):
    """Print a message with a big border."""
    border = "=" * (len(msg) + 4)
    print(f"\n{border}\n  {msg}\n{border}\n\n")


def test_forward_pass():
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 1))

    activation_sharding_num_cores = 7
    batch_size = 32
    dim = 7168
    hidden_dim = 18432

    weight = ttnn.from_torch(
        torch.rand((1, 1, hidden_dim, dim)),
        dtype=ttnn.bfloat4_b,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        memory_config=dram_sharded_weight_config(
            hidden_dim,
            dim,
            mesh_device.dram_grid_size(),
        ),
        mesh_mapper=ttnn.ttnn.shard_tensor_to_mesh_mapper(mesh_device, 0),
    )
    
    inner_memory_config = ttnn.MemoryConfig(
        memory_layout=ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        buffer_type=ttnn.BufferType.L1,
        shard_spec=ttnn.ShardSpec(
            grid=ttnn.num_cores_to_corerangeset(
                activation_sharding_num_cores, (mesh_device.core_grid.x, mesh_device.core_grid.y)
            ),
            shard_shape=(
                ttnn.core.roundup(batch_size, TILE_SIZE),
                ttnn.core.roundup(ttnn.core.divup(hidden_dim, activation_sharding_num_cores), TILE_SIZE),
            ),
            shard_orientation=ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )
    
    tt_input = ttnn.from_torch(
        torch.randn(1, 1, batch_size, hidden_dim),
        device=mesh_device,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
        dtype=ttnn.bfloat16,
        memory_config=inner_memory_config,
        layout=ttnn.TILE_LAYOUT,
    )
    
    big_print("tt_output = ttnn.linear(tt_input, weight, ...)")
    tt_output = ttnn.linear(
        tt_input,
        input_tensor_b=weight,
        memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG,
        program_config=ttnn.MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig(
            in0_block_w=1,
            per_core_M=1,
            per_core_N=ttnn.core.divup(dim, TILE_SIZE * activation_sharding_num_cores),
            fused_activation=None,
        ),
        compute_kernel_config=COMPUTE_KERNEL_CONFIG_LOFI,
    )
    report_tensor(tt_output, "tt_output")
    big_print("FINISHED EXECUTION")


test_forward_pass()
```

**Comment by yugaoTT** [comment]
a potential fix is here: https://github.com/tenstorrent/tt-metal/pull/24714
currently running pipelines to make sure all t3k and APC passes. 

The problem is it missing a check to make sure inner dim is sharded evenly (ie, all cores get the same width)

it will err out now since the passed in shard spec is not valid.

I made an extra fix that allows you to use different num cores for input and output, so that you can have the output on a evenly sharded core ranges


for example,
```
def test_forward_pass():
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 1))

    activation_sharding_num_cores = 8 #### change to 8 so that each core gets 72 tiles
    output_sharding_num_cores = 7 #### change to 7 so that each core gets 32 tiles
    batch_size = 32
    dim = 7168
    hidden_dim = 18432

    weight = ttnn.from_torch(
        torch.rand((1, 1, hidden_dim, dim)),
        dtype=ttnn.bfloat4_b,
        layout=ttnn.TILE_LAYOUT,
        device=mesh_device,
        memory_config=dram_sharded_weight_config(
            hidden_dim,
            dim,
            mesh_device.dram_grid_size(),
        ),
        mesh_mapper=ttnn.ttnn.shard_tensor_to_mesh_mapper(mesh_device, 0),
    )
    
    inner_memory_config = ttnn.MemoryConfig(
        memory_layout=ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        buffer_type=ttnn.BufferType.L1,
        shard_spec=ttnn.ShardSpec(
            grid=ttnn.num_cores_to_corerangeset(
                activation_sharding_num_cores, (mesh_device.core_grid.x, mesh_device.core_grid.y)
            ),
            shard_shape=(
                ttnn.core.roundup(batch_size, TILE_SIZE),
                ttnn.core.roundup(ttnn.core.divup(hidden_dim, activation_sharding_num_cores), TILE_SIZE),
            ),
            shard_orientation=ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )
    
    tt_input = ttnn.from_torch(
        torch.randn(1, 1, batch_size, hidden_dim),
        device=mesh_device,
        mesh_mapper=ttnn.ReplicateTensorToMesh(mesh_device),
        dtype=ttnn.bfloat16,
        memory_config=inner_memory_config,
        layout=ttnn.TILE_LAYOUT,
    )
    
    big_print("tt_output = ttnn.linear(tt_input, weight, ...)")
    tt_output = ttnn.linear(
        tt_input,
        input_tensor_b=weight,
        memory_config=ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG,
        program_config=ttnn.MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig(
            in0_block_w=1,
            per_core_M=1,
            per_core_N=ttnn.core.divup(dim, TILE_SIZE * output_sharding_num_cores),
            fused_activation=None,
        ),
        compute_kernel_config=COMPUTE_KERNEL_CONFIG_LOFI,
    )
    report_tensor(tt_output, "tt_output")
    big_print("FINISHED EXECUTION")


test_forward_pass()
```

**Comment by yugaoTT** [comment]
fix has been merged, if the commit is stabilized in main, we should be able to close
